cyclonedds/idl/_builder.py

```diff
@@ -261,7 +261,7 @@ def _machine_struct(cls, struct: Type[IdlStruct]) -> Tuple[Machine, bool]:
 
                     mutablemembers.append(MutableMember(
                         name=name,
-                        key=not keylist or name in keylist,
+                        key=keylist and name in keylist,
```

Comment:
The new code seems totally reasonable, but I wonder why it read `not keylist ...` before. It smells of marking all fields of S as a key of T in
```
struct S { long a, b; };
struct T { @key S f; };
```

IINM in such a case the TypeObject should not mark S.a, S.b as keys, and instead the serializer should take the context into account. And that would mean the new code is correct.

Am I correct in my understanding?

Reply:
Yes, that's correct, this is handled by the serializer. Members of a nested aggregated type that is the type of a key member of the top-level (or any higher-level) struct, are not marked as key, also because the same type could also be used in the same top-level type for a non-key member.